### PR TITLE
[KMSv2] GRPC Service with in-memory KMS Service

### LINKED ISSUE #112859: [KMSv2] Add metrics and tracing in `k8s.io/kms` for plugins
1. Key generation for local KEK
2. Rotation based on counter; also rotate once a week even if the counter does not hit 2 million?
3. Encryption using AES-GCM
4. Metrics
5. Logging
6. Externalized encryption of local KEK

ref: https://github.com/aramase/kubernetes-kms/blob/kms-v2alpha1/pkg/kek/kek.go

The implementation would be added to the [staging/kms](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/kms) directory.

### PR DESCRIPTION
**Pull request by ibihim**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This is a smaller version of #112898, it only contains a GRPC Service and an in-memory kms.Service.

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/112859 partially.

Fixes #

#### Special notes for your reviewer:

This is a subset of #112898 with slight changes to accommodate discussions from #sig-auth-kms-dev on kubernetes Slack.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Review by aramase — COMMENTED** [review]
/ok-to-test

File: staging/src/k8s.io/kms/encryption/aesgcm.go [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+	spaceUsage uint64 = 1 << 35
+
+	// keySize is the key size in bytes
+	keySize = 128 / 8
+)
+
+var (
+	// week was picked by during discussions of KMSv2 development.
+	week = time.Hour * 24 * 7
+	skew = time.Minute
```
Comment (enj) [inline staging/src/k8s.io/kms/encryption/aesgcm.go:49]: Not sure why we have this?

File: staging/src/k8s.io/kms/encryption/aesgcm_test.go [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+	t.Parallel()
+	ctx := context.Background()
+	aesgcm, err := NewAESGCM()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	plaintext := []byte("lorem ipsum")
+	t.Run("should be able to encrypt and decrypt from scratch", func(t *testing.T) {
+		ciphertext, err := aesgcm.Encrypt(ctx, plaintext)
```
Comment (enj) [inline staging/src/k8s.io/kms/encryption/aesgcm_test.go:36]: Probably should check that `ciphertext != plaintext`

File: staging/src/k8s.io/kms/encryption/aesgcm_test.go [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		if _, err := aesgcm.Encrypt(ctx, plaintext); err != ErrKeyExpired {
+			t.Error("key should be expired by counter")
+		}
+	})
+
+	t.Run("should be able to decrypt with unknown key, should not be able to encrypt", func(t *testing.T) {
```
Comment (enj) [inline staging/src/k8s.io/kms/encryption/aesgcm_test.go:124]: It is unclear to me why we would need this unknown key concept.

**Review by enj — CHANGES_REQUESTED** [review]

File: staging/src/k8s.io/kms/encryption/aesgcm.go (hunk 1 above)
Comment (ibihim) [reply staging/src/k8s.io/kms/encryption/aesgcm.go:49]: I added this in certain parts of the code, where I would like to ensure that we are definitely within the time frame we want.

Removed it.

File: staging/src/k8s.io/kms/encryption/aesgcm.go [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+// FromKey initializes a cipher.AEAD from bytes, which is marked as expired.
+// It is unknown how often it is already used, so we need to assume it is unsafe.
+func FromKey(key []byte) (*AESGCM, error) {
+	return newAESGCM(
+		key,
+		maxUsage,
+		time.Now().Add(-week).Add(-skew),
+	)
+}
```
[the remark this answers is not in the document]
Comment (ibihim) [reply staging/src/k8s.io/kms/encryption/aesgcm.go:104]: This is currently not necessary. The intention is to initialize unknown local KEKs (KEKs created by other plugins) in the future.
As we don't know how often it was used by the other plugin, we start of with values that make it unusable for the current plugin.

Removed it.

File: staging/src/k8s.io/kms/encryption/aesgcm_test.go [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+func TestAESGCM(t *testing.T) {
+	t.Parallel()
+	ctx := context.Background()
+	aesgcm, err := NewAESGCM()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	plaintext := []byte("lorem ipsum")
+	t.Run("should be able to encrypt and decrypt from scratch", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (ibihim) [reply staging/src/k8s.io/kms/encryption/aesgcm_test.go:35]: yes, let me add it.

File: staging/src/k8s.io/kms/encryption/memory.go [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+	"k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2"
+)
+
+// InMemory is a kmsv2.Service wrapper around AESGCM. It can be used for testing
+// or experimenting. It isn't meant to be used on production as the key is only
+// in memory.
+type InMemory struct {
+	keyID  string
+	cipher *AESGCM
```
[the remark this answers is not in the document]
Comment (ibihim) [reply staging/src/k8s.io/kms/encryption/memory.go:33]: 💯 %

File: staging/src/k8s.io/kms/encryption/memory.go [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+var (
+	_ kmsv2.Service = (*InMemory)(nil)
```
[the remark this answers is not in the document]
Comment (ibihim) [reply staging/src/k8s.io/kms/encryption/memory.go:37]: Ok, done. I put the interface into `k8s.io/kms/service`.
I could move the`memory.go` into the `service` package as well... and then keep `grpc_service` in there as well or move it into a `server` package.

WDYT?

File: staging/src/k8s.io/kms/encryption/memory.go [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+// Status returns the KeyID (UUID), the current API version and a health status
+// message. The health status returns "ok" if everything is working, and returns
+// a more detailed version, if there are limitations on its usage.
+func (m *InMemory) Status(ctx context.Context) (*kmsv2.StatusResponse, error) {
+	health := "ok"
+	if !m.cipher.IsValid() {
+		health = "decrypt: ok, encrypt: ErrKeyExpired"
+	}
```
[the remark this answers is not in the document]
Comment (ibihim) [reply staging/src/k8s.io/kms/encryption/memory.go:73]: So you are suggesting to not offer a `IsValid` function in the `kms/service.Service interface`, but to rely on an err check? Sure, seems legit.

File: staging/src/k8s.io/kms/service/grpc_service.go [hunk 9]
```diff
@@ -0,0 +1,137 @@
+/*
```
[the remark this answers is not in the document]
Comment (ibihim) [reply staging/src/k8s.io/kms/service/grpc_service.go:1]: Sorry.

File: staging/src/k8s.io/kms/encryption/aesgcm.go [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+
+	_ = atomic.AddUint32(&c.counter, 1)
+
+	return c.transformer.TransformToStorage(ctx, plaintext, value.DefaultContext{})
+}
+
+// Decrypt decrypts a ciphertext. The nonce is assumed to be prepended.
+// Therefore any change to the standard nonceSize is a breaking change.
+func (k *AESGCM) Decrypt(ctx context.Context, ciphertext []byte) ([]byte, error) {
+	plaintext, _, err := k.transformer.TransformFromStorage(ctx, ciphertext, value.DefaultContext{})
```
[the remark this answers is not in the document]
Comment (ibihim) [reply staging/src/k8s.io/kms/encryption/aesgcm.go:129]: What do you have in mind? I wouldn't put anything into the additional data of AEAD.
We could consider adding the ... keyID?

File: staging/src/k8s.io/kms/go.mod [hunk 11]
```diff
@@ -6,17 +6,37 @@ go 1.19
...
 	google.golang.org/protobuf v1.28.1 // indirect
+	k8s.io/apimachinery v0.0.0 // indirect
+	k8s.io/component-base v0.0.0 // indirect
+	k8s.io/utils v0.0.0-20221107191617-1a15be271d1d // indirect
 )
 
-replace k8s.io/kms => ../kms
+replace (
+	k8s.io/apimachinery => ../apimachinery
+	k8s.io/apiserver => ../apiserver
```
[the remark this answers is not in the document]
Comment (ibihim) [reply staging/src/k8s.io/kms/go.mod:39]: Yes, sure. We want to use it in apiserver. I will copy as you suggested.

File: staging/src/k8s.io/kms/service/grpc_service_test.go [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+		if _, err := uuid.Parse(status.KeyId); err != nil {
+			t.Error(err)
+		}
+		if status.Version != "v2alpha1" {
+			t.Errorf("want %q, have: %q", "v2alpha1", status.Version)
+		}
+	})
+}
+
+func newClient(address string) (kmsapi.KeyManagementServiceClient, func() error, error) {
```
[the remark this answers is not in the document]
Comment (ibihim) [reply staging/src/k8s.io/kms/service/grpc_service_test.go:111]: Hm, I prefer to bubble the error up and fail there, so it is not dependent on `*testing.T`, but I can hand it down.

File: staging/src/k8s.io/kms/encryption/aesgcm.go [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+// The maximum length of bytes that can be encrypted with one key and nonce is
+// ((2^36)-256) bytes.
+// The amount of times a key with different random nonces can be used, until the
+// probability of re-using a nonce becomes to high is 2^32.
+type AESGCM struct {
+	key     []byte
+	counter uint32
+	expiry  time.Time
+
+	transformer value.Transformer
```
[the remark this answers is not in the document]
Comment (ibihim) [reply staging/src/k8s.io/kms/encryption/aesgcm.go:69]: As I thought of it, AESGCM is a wrapper around the AES GCM Transformer. It is also very specific to the Cryptographic boundaries of AES GCM. AES CBC has different boundaries and would need its own wrapper that handles its boundaries.
In the next PR I would then introduce the `local KEK`, which then is actually a `func(Transformer) Transformer`, as it is an interceptor to the upstream Transformer. This implementation could use other encryption transformer like AES CBC or ChaCha20-Poly1305.

**Comment by dims** [comment]
If you still need this PR then please rebase, if not, please close the PR

**Comment by ibihim** [comment]
With regards to the design, I would consider something that looks like so:

High Level:

```mermaid
flowchart LR
  grpcService["grpc service interface"]
  grpcServer["grpc server"]
  kmsService["kms service interface"]

  grpcService -->| implemented by | grpcServer
  grpcServer -->| uses | kmsService
```

Within KMS Plugin:

```mermaid
flowchart LR
  kmsService["kms service interface"]
  kmsRemote["cloud provider kms client"]
  hsm["HSM"]
  kmsLocal["local kms service (middleware)"]
  
  kmsService -->| can be | kmsRemote
  kmsService -->| can be | hsm
  kmsService -->| can be | kmsLocal
  kmsLocal --> | uses | kmsService
```

Local KMS Middleware High Level:

```mermaid
flowchart LR
  kmsLocal["kms local service"]
  kmsService["kms service interface"]
  transformers["List of transformers"]
  store["transformer store"]

  kmsLocal -->| has | kmsService
  kmsLocal -->| implements | kmsService
  kmsLocal -->| has | transformers
  kmsLocal -->| has | store
  transformers --> | stored at | store
```

With regards to cloud provider kms client:

```mermaid
flowchart LR
  kmsLocal["kms local service"]
  kmsService["kms service interface"]
  kmsRemote["cloud provider kms client"]

  kmsLocal -->| uses for local KEK | kmsService
  kmsLocal --> | is a | kmsService
  kmsService --> | is most probably | kmsRemote
```

With regards to the transformers:

```mermaid
flowchart LR
  kmsLocal["kms local service"]
  transformer["current transformer"]
  transformers["List of transformers"]
  store["transformer store"]

  kmsLocal -->| encrypts data with | transformer
  kmsLocal -->| decrypts data with | transformers
  kmsLocal -->| has | store
  transformer --> | part of | transformers
  transformers --> | stored at | store
```

`kms local service` is used as a middleware for a `cloud provider's kms client`, like AWS, GCP, Azure ...
I would've created an constructor for `kms local service` that expects a `args` and a `kms service`, while also returning a `kms service`.
I didn't think of having a dedicated middleware type (`func kms service : kms service`) as a property of the `kms local service` itself.
If you have suggestions, I am happy to discuss that in detail.

As you suggested, the GRPC part would've been the right size for this PR. I can trim down the PR to make it smaller.

I thought it would've been a good idea to also add a transformer with a shim (in memory), to let it make something fancy.

**Comment by ibihim** [comment]
Closed in favor of a smaller PR: #114678

**Pull request closed without merging**